File: src/main/java/com/telecom/utils/ScreenshotUtils.java
package com.telecom.utils;

import com.telecom.config.SIMToolkitConfig;
import io.appium.java_client.AppiumDriver;
import org.apache.commons.io.FileUtils;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;

import java.io.File;
import java.lang.reflect.Method;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@SuppressWarnings("unused")
public class ScreenshotUtils {

    private AppiumDriver driver;
    private volatile int screenshotCounter = 0;
    private final List<String> capturedScreenshots = Collections.synchronizedList(new ArrayList<>());
    private final List<String> screenshotDescriptions = Collections.synchronizedList(new ArrayList<>());
    private final List<Long> screenshotTimestamps = Collections.synchronizedList(new ArrayList<>());
    private Date testStartTime;
    private Date testEndTime;
    
    // Constants
    private static final String REPORT_DIR = "test-output/comprehensive-reports/";
    private static final Pattern TIMESTAMP_PATTERN = 
        Pattern.compile("_(\\d{8})_(\\d{6})\\.png$");
    
    // Thread-safe date formatters
    private static final ThreadLocal<SimpleDateFormat> TIMESTAMP_FORMATTER = 
        ThreadLocal.withInitial(() -> new SimpleDateFormat("yyyyMMdd_HHmmss"));
    private static final ThreadLocal<SimpleDateFormat> DISPLAY_FORMATTER = 
        ThreadLocal.withInitial(() -> new SimpleDateFormat("yyyy-MM-dd HH:mm:ss"));
    private static final ThreadLocal<SimpleDateFormat> FILE_DATE_FORMATTER = 
        ThreadLocal.withInitial(() -> new SimpleDateFormat("yyyyMMdd"));

    public ScreenshotUtils(AppiumDriver driver) {
        this.driver = driver;
        createScreenshotDir();
        loadExistingScreenshots();
    }
    
    /**
     * Set test start time - call this at the beginning of test
     */
    public void setTestStartTime() {
        this.testStartTime = new Date();
        System.out.println("⏰ Test start time set: " + testStartTime);
    }
    
    /**
     * Set test end time - call this at the end of test
     */
    public void setTestEndTime() {
        this.testEndTime = new Date();
        System.out.println("⏰ Test end time set: " + testEndTime);
    }
    
    public Date getTestStartTime() {
        return testStartTime;
    }
    
    public Date getTestEndTime() {
        return testEndTime;
    }

    /**
     * Create screenshot directory if it doesn't exist
     */
    private void createScreenshotDir() {
        File dir = new File(SIMToolkitConfig.SCREENSHOT_DIR);
        if (!dir.exists()) {
            boolean created = dir.mkdirs();
            if (created) {
                System.out.println("📁 Created screenshot directory: " + SIMToolkitConfig.SCREENSHOT_DIR);
            } else {
                System.err.println("❌ Failed to create screenshot directory: " + SIMToolkitConfig.SCREENSHOT_DIR);
            }
        }
    }

    /**
     *  Load existing screenshots from disk into memory
     * FIXED: Added null check for directory
     */
    private void loadExistingScreenshots() {
        File dir = new File(SIMToolkitConfig.SCREENSHOT_DIR);
        
        // FIX: Check if directory exists before listing files
        if (!dir.exists() || !dir.isDirectory()) {
            System.out.println("📁 Screenshot directory not found or is not a directory: " + SIMToolkitConfig.SCREENSHOT_DIR);
            return;
        }
        
        File[] files = dir.listFiles((d, name) -> name.endsWith(".png"));
        
        if (files != null && files.length > 0) {
            Arrays.sort(files, Comparator.comparing(File::getName));
            
            for (File file : files) {
                capturedScreenshots.add(file.getName());
                String desc = file.getName()
                    .replaceAll("^\\d+_", "")
                    .replaceAll("_\\d{8}_\\d{6}\\.png$", "")
                    .replace("_", " ");
                screenshotDescriptions.add(desc);
                
                // Extract and store timestamp
                Long timestamp = extractTimestampFromFilename(file.getName());
                screenshotTimestamps.add(timestamp);
            }
            
            screenshotCounter = files.length;
            System.out.println("📂 Loaded " + files.length + " existing screenshots from disk");
        } else {
            System.out.println("📂 No existing screenshots found in directory: " + SIMToolkitConfig.SCREENSHOT_DIR);
        }
    }
    
    /**
     * Extract timestamp from filename (e.g., 001_launch_20241220_143022.png)
     */
    private Long extractTimestampFromFilename(String filename) {
        try {
            Matcher matcher = TIMESTAMP_PATTERN.matcher(filename);
            if (matcher.find()) {
                String dateStr = matcher.group(1); // yyyyMMdd
                String timeStr = matcher.group(2); // HHmmss
                
                SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss");
                Date date = sdf.parse(dateStr + "_" + timeStr);
                return date.getTime();
            }
        } catch (Exception e) {
            System.err.println("❌ Error parsing timestamp from filename: " + filename + " - " + e.getMessage());
        }
        return 0L;
    }

    /**
     * Capture screenshot with step name
     * FIXED: Added null check for driver and proper file path construction
     */
    public String captureScreenshot(String stepName) {
        if (driver == null) {
            System.err.println("❌ Driver is null, cannot capture screenshot for: " + stepName);
            return null;
        }
        
        try {
            screenshotCounter++;
            
            String timestamp = TIMESTAMP_FORMATTER.get().format(new Date());
            long currentTime = System.currentTimeMillis();
            
            // FIX: Sanitize step name to remove invalid filename characters
            String sanitizedStepName = stepName.toLowerCase()
                .replace(" ", "_")
                .replaceAll("[^a-z0-9_]", "");
            
            String fileName = String.format(
                    "%03d_%s_%s.png",
                    screenshotCounter,
                    sanitizedStepName,
                    timestamp
            );

            // FIX: Ensure directory exists before saving
            File screenshotDir = new File(SIMToolkitConfig.SCREENSHOT_DIR);
            if (!screenshotDir.exists()) {
                screenshotDir.mkdirs();
            }
            
            // FIX: Use proper file path construction
            File dest = new File(screenshotDir, fileName);
            
            File src = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
            FileUtils.copyFile(src, dest);

            capturedScreenshots.add(fileName);
            screenshotDescriptions.add(stepName);
            screenshotTimestamps.add(currentTime);

            System.out.println("📸 Screenshot saved: " + dest.getAbsolutePath());
            return dest.getAbsolutePath();

        } catch (Exception e) {
            System.err.println("❌ Screenshot failed for " + stepName + ": " + e.getMessage());
            e.printStackTrace();
            return null;
        }
    }
    
    /**
     * Get screenshots captured during the test period
     */
    private List<ScreenshotInfo> getScreenshotsInTestPeriod() {
        List<ScreenshotInfo> testScreenshots = new ArrayList<>();
        
        if (testStartTime == null || testEndTime == null) {
            System.out.println(" Test start/end time not set. Showing all screenshots.");
            // Return all if times not set
            synchronized (capturedScreenshots) {
                for (int i = 0; i < capturedScreenshots.size(); i++) {
                    testScreenshots.add(new ScreenshotInfo(
                        capturedScreenshots.get(i),
                        screenshotDescriptions.get(i),
                        screenshotTimestamps.get(i)
                    ));
                }
            }
            return testScreenshots;
        }
        
        long startTime = testStartTime.getTime();
        long endTime = testEndTime.getTime();
        
        System.out.println("🔍 Filtering screenshots from " + testStartTime + " to " + testEndTime);
        
        synchronized (capturedScreenshots) {
            for (int i = 0; i < capturedScreenshots.size(); i++) {
                long screenshotTime = screenshotTimestamps.get(i);
                
                // Check if screenshot was taken during test period
                // Adding 2 second buffer for timing differences
                if (screenshotTime >= (startTime - 2000) && screenshotTime <= (endTime + 2000)) {
                    testScreenshots.add(new ScreenshotInfo(
                        capturedScreenshots.get(i),
                        screenshotDescriptions.get(i),
                        screenshotTimestamps.get(i)
                    ));
                }
            }
        }
        
        System.out.println("📊 Found " + testScreenshots.size() + " screenshots in test period");
        return testScreenshots;
    }
    
    /**
     * Helper class to store screenshot information
     */
    private static class ScreenshotInfo {
        String filename;
        String description;
        long timestamp;
        
        ScreenshotInfo(String filename, String description, long timestamp) {
            this.filename = filename;
            this.description = description;
            this.timestamp = timestamp;
        }
        
        String getFormattedTime() {
            return DISPLAY_FORMATTER.get().format(new Date(timestamp));
        }
        
        /**
         * Get relative path for HTML report
         * FIXED: Proper path calculation
         */
        String getRelativeImagePath() {
            try {
                File screenshotFile = new File(SIMToolkitConfig.SCREENSHOT_DIR, filename);
                File reportDir = new File(REPORT_DIR);
                
                if (!reportDir.exists()) {
                    return "file:///" + screenshotFile.getAbsolutePath().replace("\\", "/");
                }
                
                // Calculate relative path
                String screenshotPath = screenshotFile.getCanonicalPath();
                String reportPath = reportDir.getCanonicalPath();
                
                if (screenshotPath.startsWith(reportPath)) {
                    return screenshotPath.substring(reportPath.length()).replace("\\", "/");
                }
                
                // Fallback: Use relative path if in same parent directory
                return "../" + new File(SIMToolkitConfig.SCREENSHOT_DIR).getName() + "/" + filename;
            } catch (Exception e) {
                System.err.println("❌ Error calculating relative path for: " + filename);
                return filename; // Fallback
            }
        }
    }

    /**
     * Generate HTML screenshot report
     * FIXED: Report directory constant and image path calculation
     */
    public void generateScreenshotReport() {
        try {
            // FIX: Use constant for report directory
            File reportDir = new File(REPORT_DIR);
            if (!reportDir.exists()) {
                reportDir.mkdirs();
                System.out.println("📁 Created report directory: " + REPORT_DIR);
            }

            // Get the device number like other reports do
            String dialingNumber = System.getProperty("aPartyNumber");
            if (dialingNumber == null || dialingNumber.isEmpty()) {
                dialingNumber = "unknown";
            }

            String timestamp = TIMESTAMP_FORMATTER.get().format(new Date());
            // FIX: Use proper file path construction
            String reportPath = new File(reportDir, 
                "Screenshot_Report_" + dialingNumber + "_" + timestamp + ".html").getAbsolutePath();
            
            // Get screenshots captured during test period
            List<ScreenshotInfo> testScreenshots = getScreenshotsInTestPeriod();

            StringBuilder html = new StringBuilder();
            html.append("<!DOCTYPE html><html><head>")
                .append("<meta charset='UTF-8'>")
                .append("<meta name='viewport' content='width=device-width, initial-scale=1.0'>")
                .append("<title>Vi SIM Toolkit Screenshot Report - ").append(dialingNumber).append("</title>")
                .append("<style>")
                .append("body { font-family: 'Segoe UI', Arial, sans-serif; background: #f5f7fa; padding: 20px; margin: 0; }")
                .append("h1 { color: #2c3e50; border-bottom: 3px solid #3498db; padding-bottom: 10px; }")
                .append(".info-box { background: white; padding: 20px; border-radius: 8px; margin: 15px 0; ")
                .append("box-shadow: 0 2px 4px rgba(0,0,0,0.1); }")
                .append(".info-box h2 { color: #2980b9; margin-top: 0; }")
                .append(".info-grid { display: grid; grid-template-columns: repeat(auto-fit, minmax(300px, 1fr)); ")
                .append("gap: 15px; margin-top: 15px; }")
                .append(".info-item { background: #ecf0f1; padding: 15px; border-radius: 6px; }")
                .append(".info-item strong { display: block; color: #34495e; margin-bottom: 5px; }")
                .append(".screenshot-box { background: white; padding: 20px; margin: 20px 0; ")
                .append("border-radius: 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.1); }")
                .append(".screenshot-box h3 { color: #2980b9; margin-top: 0; border-bottom: 2px solid #ecf0f1; padding-bottom: 8px; }")
                .append(".screenshot-meta { color: #7f8c8d; font-size: 14px; margin: 8px 0; }")
                .append(".screenshot-box img { max-width: 400px; height: auto; border: 2px solid #bdc3c7; ")
                .append("border-radius: 4px; margin-top: 10px; box-shadow: 0 2px 8px rgba(0,0,0,0.15); ")
                .append("transition: transform 0.2s; }")
                .append(".screenshot-box img:hover { transform: scale(1.02); }")
                .append(".screenshot-box code { background: #ecf0f1; padding: 4px 8px; border-radius: 4px; ")
                .append("color: #e74c3c; font-size: 12px; }")
                .append(".footer { text-align: center; color: #7f8c8d; margin-top: 30px; padding: 20px; ")
                .append("border-top: 1px solid #ecf0f1; }")
                .append(".no-screenshots { text-align: center; padding: 40px; color: #7f8c8d; ")
                .append("font-style: italic; background: #ecf0f1; border-radius: 8px; margin: 20px 0; }")
                .append(".image-container { position: relative; }")
                .append(".image-not-found { color: #e74c3c; background: #ffeaea; padding: 10px; ")
                .append("border-radius: 4px; margin-top: 10px; }")
                .append("</style></head><body>");

            html.append("<h1> Vi SIM Toolkit Screenshot Report - ").append(dialingNumber).append("</h1>");
            
            // Test Information Section
            html.append("<div class='info-box'>")
                .append("<h2>Test Information</h2>")
                .append("<div class='info-grid'>");
            
            // Add device number to report
            html.append("<div class='info-item'>")
                .append("<strong>Device Number</strong>")
                .append("<span>").append(dialingNumber.equals("unknown") ? "Not Specified" : dialingNumber).append("</span>")
                .append("</div>");
            
            if (testStartTime != null) {
                html.append("<div class='info-item'>")
                    .append("<strong>Test Start Time</strong>")
                    .append("<span>").append(DISPLAY_FORMATTER.get().format(testStartTime)).append("</span>")
                    .append("</div>");
            }
            
            if (testEndTime != null) {
                html.append("<div class='info-item'>")
                    .append("<strong>Test End Time</strong>")
                    .append("<span>").append(DISPLAY_FORMATTER.get().format(testEndTime)).append("</span>")
                    .append("</div>");
            }
            
            html.append("<div class='info-item'>")
                .append("<strong>📸 Total Screenshots in Test Period</strong>")
                .append("<span>").append(testScreenshots.size()).append("</span>")
                .append("</div>")
//                .append("<div class='info-item'>")
//                .append("<strong>📂 Screenshot Directory</strong>")
//                .append("<span><code>").append(SIMToolkitConfig.SCREENSHOT_DIR).append("</code></span>")
//                .append("</div>")
//                .append("<div class='info-item'>")
//                .append("<strong>📊 Report Location</strong>")
//                .append("<span><code>").append(reportPath).append("</code></span>")
//                .append("</div>")
                .append("</div></div>");
            
            // Screenshots Section
            html.append("<div class='info-box'>")
                .append("<h2>Screenshots Captured During Test</h2>");
            
            if (testScreenshots.isEmpty()) {
                html.append("<div class='no-screenshots'>")
                    .append("<p> No screenshots found for the specified test period</p>")
                    .append("<p>Test Period: ");
                
                if (testStartTime != null) {
                    html.append(DISPLAY_FORMATTER.get().format(testStartTime));
                }
                html.append(" to ");
                if (testEndTime != null) {
                    html.append(DISPLAY_FORMATTER.get().format(testEndTime));
                }
                html.append("</p></div>");
            } else {
                for (int i = 0; i < testScreenshots.size(); i++) {
                    ScreenshotInfo info = testScreenshots.get(i);
                    // FIX: Use calculated relative path
                    String imagePath = info.getRelativeImagePath();
                    
                    html.append("<div class='screenshot-box'>")
                        .append("<h3>").append(i + 1).append(". ").append(info.description).append("</h3>")
                        .append("<div class='screenshot-meta'>")
                        .append("<strong>File:</strong> <code>").append(info.filename).append("</code><br>")
                        .append("<strong>Captured at:</strong> ").append(info.getFormattedTime())
                        .append("</div>")
                        .append("<div class='image-container'>")
                        .append("<img src=\"").append(imagePath).append("\" ")
                        .append("alt='").append(info.description).append("' ")
                        .append("onerror=\"this.style.display='none';")
                        .append("this.parentElement.innerHTML+='<div class=\\'image-not-found\\'>❌ Image not found: ")
                        .append(info.filename).append("</div>';\">")
                        .append("</div>")
                        .append("</div>");
                }
            }
            
            html.append("</div>");

            html.append("<div class='footer'>")
                .append("<p>📂 Report saved in: ").append(REPORT_DIR).append("</p>")
                .append("<p>Generated by Vi SIM Toolkit Automation Framework</p>")
                .append("<p>Report generated: ").append(DISPLAY_FORMATTER.get().format(new Date())).append("</p>")
                .append("</div>")
                .append("</body></html>");

            FileUtils.writeStringToFile(new File(reportPath), html.toString(), "UTF-8");
            System.out.println(" Screenshot report saved: " + reportPath);
            System.out.println("📊 Screenshots in report: " + testScreenshots.size());

        } catch (Exception e) {
            System.err.println("❌ Screenshot report generation failed: " + e.getMessage());
            e.printStackTrace();
        }
    }

    public Map<String, Boolean> verifyRequiredScreenshots(SIMToolkitConfig.SIMType simType) {
        Map<String, Boolean> results = new LinkedHashMap<>();
        List<String> required = getRequiredScreenshots(simType);

        File dir = new File(SIMToolkitConfig.SCREENSHOT_DIR);
        if (!dir.exists() || !dir.isDirectory()) {
            // Directory doesn't exist, mark all as false
            for (String req : required) {
                results.put(req, false);
            }
            return results;
        }
        
        File[] actualFiles = dir.listFiles((d, n) -> n.endsWith(".png"));

        for (String req : required) {
            boolean found = false;
            String key = req.toLowerCase().replace(" ", "_");

            if (actualFiles != null) {
                for (File f : actualFiles) {
                    if (f.getName().toLowerCase().contains(key)) {
                        found = true;
                        break;
                    }
                }
            }
            results.put(req, found);
        }
        return results;
    }

    public void clearScreenshots() {
        try {
            File dir = new File(SIMToolkitConfig.SCREENSHOT_DIR);
            if (dir.exists() && dir.isDirectory()) {
                FileUtils.cleanDirectory(dir);
                System.out.println("🧹 Cleared screenshot directory: " + SIMToolkitConfig.SCREENSHOT_DIR);
            }
            
            synchronized (capturedScreenshots) {
                capturedScreenshots.clear();
                screenshotDescriptions.clear();
                screenshotTimestamps.clear();
            }
            screenshotCounter = 0;
            
        } catch (Exception e) {
            System.err.println("❌ Clear screenshots failed: " + e.getMessage());
        }
    }

    public void printScreenshotSummary() {
        int size;
        synchronized (capturedScreenshots) {
            size = capturedScreenshots.size();
        }
        
        System.out.println("   Total screenshots: " + size);
        if (size == 0) {
            System.out.println("    No screenshots found!");
        } else {
            synchronized (capturedScreenshots) {
                for (int i = 0; i < Math.min(size, 10); i++) {
                    System.out.println("   " + (i + 1) + ". " + screenshotDescriptions.get(i));
                }
                if (size > 10) {
                    System.out.println("   ... and " + (size - 10) + " more");
                }
            }
        }
    }

    private List<String> getRequiredScreenshots(SIMToolkitConfig.SIMType simType) {
        List<String> list = new ArrayList<>();
        list.add("SIM Toolkit Launch");
        if (simType != SIMToolkitConfig.SIMType.SINGLE_SIM) {
            list.add("SIM Selection Screen");
        }
        list.add("Vi Menu Home");
        list.add("Flash Option");
        list.add("Roaming Menu");
        list.add("Vodafone IN");
        list.add("International");
        return list;
    }
    
    public int getScreenshotCount() {
        synchronized (capturedScreenshots) {
            return capturedScreenshots.size();
        }
    }
    
    /**
     * Optional: Get list of all screenshot files with full paths
     */
    public List<String> getAllScreenshotPaths() {
        List<String> paths = new ArrayList<>();
        synchronized (capturedScreenshots) {
            for (String filename : capturedScreenshots) {
                paths.add(new File(SIMToolkitConfig.SCREENSHOT_DIR, filename).getAbsolutePath());
            }
        }
        return paths;
    }
}

File: test/utils/ScreenshotUtils.ts
import fs from 'fs';
import path from 'path';
import type { AndroidDriver } from '../types/driver';
import { SCREENSHOT_DIR, REPORT_DIR, SIMType } from '../config/SIMToolkitConfig';

interface ScreenshotInfo {
  filename: string;
  description: string;
  timestamp: number;
}

const TIMESTAMP_PATTERN = /_(\d{8})_(\d{6})\.png$/;

/**  com.telecom.utils.ScreenshotUtils */
export class ScreenshotUtils {
  private counter = 0;
  private readonly filenames: string[] = [];
  private readonly descriptions: string[] = [];
  private readonly timestamps: number[] = [];
  private testStartTime?: Date;
  private testEndTime?: Date;

  constructor(private readonly driver: AndroidDriver) {
    if (!fs.existsSync(SCREENSHOT_DIR)) {
      fs.mkdirSync(SCREENSHOT_DIR, { recursive: true });
    }
    this.loadExistingScreenshots();
  }

  setTestStartTime(): void {
    this.testStartTime = new Date();
    console.log(`Test start time set: ${this.testStartTime.toISOString()}`);
  }

  setTestEndTime(): void {
    this.testEndTime = new Date();
    console.log(`Test end time set: ${this.testEndTime.toISOString()}`);
  }

  getTestStartTime(): Date | undefined {
    return this.testStartTime;
  }

  getTestEndTime(): Date | undefined {
    return this.testEndTime;
  }

  async clearScreenshots(): Promise<void> {
    this.counter = 0;
    this.filenames.length = 0;
    this.descriptions.length = 0;
    this.timestamps.length = 0;
    if (fs.existsSync(SCREENSHOT_DIR)) {
      for (const f of fs.readdirSync(SCREENSHOT_DIR)) {
        if (f.endsWith('.png')) fs.unlinkSync(path.join(SCREENSHOT_DIR, f));
      }
    }
    console.log(`Cleared screenshot directory: ${SCREENSHOT_DIR}`);
  }

  async captureScreenshot(stepName: string): Promise<string | null> {
    try {
      this.counter += 1;
      const stamp = formatTimestamp(new Date());
      const safe = stepName
        .toLowerCase()
        .replace(/\s+/g, '_')
        .replace(/[^a-z0-9_]/g, '');

      const fileName = `${String(this.counter).padStart(3, '0')}_${safe}_${stamp}.png`;
      const filePath = path.join(SCREENSHOT_DIR, fileName);
      const now = Date.now();

      const png = await this.driver.takeScreenshot();
      fs.writeFileSync(filePath, png, 'base64');

      this.filenames.push(fileName);
      this.descriptions.push(stepName);
      this.timestamps.push(now);

      console.log(`Screenshot saved: ${filePath}`);
      return filePath;
    } catch (e) {
      console.error(`Screenshot failed for ${stepName}: ${(e as Error).message}`);
      return null;
    }
  }

  async verifyRequiredScreenshots(simType: SIMType): Promise<Record<string, boolean>> {
    const required = ['SIM Toolkit Launch'];
    if (simType !== SIMType.SINGLE_SIM) required.push('SIM Selection Screen');
    required.push('Vi Menu Home', 'Flash Option', 'Roaming Menu', 'Vodafone IN', 'International');

    const files = fs.existsSync(SCREENSHOT_DIR)
      ? fs.readdirSync(SCREENSHOT_DIR).filter(f => f.endsWith('.png'))
      : [];

    const result: Record<string, boolean> = {};
    for (const req of required) {
      const key = req.toLowerCase().replace(/\s+/g, '_');
      result[req] = files.some(f => f.toLowerCase().includes(key));
    }
    return result;
  }

  /** Generate HTML screenshot report —  generateScreenshotReport() */
  async generateScreenshotReport(): Promise<string | null> {
    try {
      if (!fs.existsSync(REPORT_DIR)) fs.mkdirSync(REPORT_DIR, { recursive: true });

      const dialingNumber =
        process.env.APARTY_NUMBER || process.env.A_PARTY_NUMBER || 'unknown';
      const stamp = formatTimestamp(new Date());
      const reportPath = path.join(
        REPORT_DIR,
        `Screenshot_Report_${dialingNumber}_${stamp}.html`
      );

      const shots = this.getScreenshotsInTestPeriod();
      const html = buildScreenshotHtml(dialingNumber, shots, this.testStartTime, this.testEndTime);
      fs.writeFileSync(reportPath, html, 'utf8');

      console.log(`Screenshot report saved: ${reportPath}`);
      console.log(`Screenshots in report: ${shots.length}`);
      return reportPath;
    } catch (e) {
      console.error(`Screenshot report generation failed: ${(e as Error).message}`);
      return null;
    }
  }

  printScreenshotSummary(): void {
    console.log(`   Total screenshots: ${this.filenames.length}`);
    if (!this.filenames.length) {
      console.log('   No screenshots found!');
      return;
    }
    for (let i = 0; i < Math.min(this.filenames.length, 10); i++) {
      console.log(`   ${i + 1}. ${this.descriptions[i]}`);
    }
    if (this.filenames.length > 10) {
      console.log(`   ... and ${this.filenames.length - 10} more`);
    }
  }

  getScreenshotCount(): number {
    return this.filenames.length;
  }

  getAllScreenshotPaths(): string[] {
    return this.filenames.map(f => path.join(SCREENSHOT_DIR, f));
  }

  private loadExistingScreenshots(): void {
    if (!fs.existsSync(SCREENSHOT_DIR)) return;
    const files = fs
      .readdirSync(SCREENSHOT_DIR)
      .filter(f => f.endsWith('.png'))
      .sort();

    for (const file of files) {
      this.filenames.push(file);
      this.descriptions.push(
        file.replace(/^\d+_/, '').replace(TIMESTAMP_PATTERN, '').replace(/_/g, ' ')
      );
      this.timestamps.push(extractTimestampFromFilename(file) || 0);
    }
    this.counter = files.length;
    if (files.length) console.log(`Loaded ${files.length} existing screenshots from disk`);
  }

  private getScreenshotsInTestPeriod(): ScreenshotInfo[] {
    const all: ScreenshotInfo[] = this.filenames.map((filename, i) => ({
      filename,
      description: this.descriptions[i] || filename,
      timestamp: this.timestamps[i] || 0
    }));

    if (!this.testStartTime || !this.testEndTime) return all;

    const start = this.testStartTime.getTime() - 2000;
    const end = this.testEndTime.getTime() + 2000;
    return all.filter(s => s.timestamp >= start && s.timestamp <= end);
  }
}

function formatTimestamp(d: Date): string {
  const p = (n: number) => String(n).padStart(2, '0');
  return `${d.getFullYear()}${p(d.getMonth() + 1)}${p(d.getDate())}_${p(d.getHours())}${p(d.getMinutes())}${p(d.getSeconds())}`;
}

function extractTimestampFromFilename(filename: string): number | null {
  const m = filename.match(TIMESTAMP_PATTERN);
  if (!m) return null;
  const d = new Date(
    parseInt(m[1].slice(0, 4), 10),
    parseInt(m[1].slice(4, 6), 10) - 1,
    parseInt(m[1].slice(6, 8), 10),
    parseInt(m[2].slice(0, 2), 10),
    parseInt(m[2].slice(2, 4), 10),
    parseInt(m[2].slice(4, 6), 10)
  );
  return isNaN(d.getTime()) ? null : d.getTime();
}

function relativeImagePath(filename: string): string {
  return `../screenshots/${filename}`;
}

function buildScreenshotHtml(
  dialingNumber: string,
  shots: ScreenshotInfo[],
  start?: Date,
  end?: Date
): string {
  const fmt = (d?: Date) => (d ? d.toLocaleString() : 'N/A');

  let body = shots
    .map(
      (s, i) => `
    <div class="screenshot-box">
      <h3>${i + 1}. ${escapeHtml(s.description)}</h3>
      <div class="screenshot-meta">
        <strong>File:</strong> <code>${escapeHtml(s.filename)}</code><br>
        <strong>Captured at:</strong> ${new Date(s.timestamp).toLocaleString()}
      </div>
      <img src="${relativeImagePath(s.filename)}" alt="${escapeHtml(s.description)}"
        onerror="this.style.display='none';this.nextElementSibling.style.display='block';">
      <div class="image-not-found" style="display:none">Image not found: ${escapeHtml(s.filename)}</div>
    </div>`
    )
    .join('');

  if (!shots.length) {
    body = `<div class="no-screenshots"><p>No screenshots found for the test period (${fmt(start)} → ${fmt(end)})</p></div>`;
  }

  return `<!DOCTYPE html><html><head><meta charset="UTF-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>Vi Screenshot Report - ${escapeHtml(dialingNumber)}</title>
<style>
body{font-family:'Segoe UI',Arial,sans-serif;background:#f5f7fa;padding:20px;margin:0}
h1{color:#2c3e50;border-bottom:3px solid #3498db;padding-bottom:10px}
.info-box,.screenshot-box{background:#fff;padding:20px;border-radius:8px;margin:15px 0;box-shadow:0 2px 4px rgba(0,0,0,.1)}
.info-grid{display:grid;grid-template-columns:repeat(auto-fit,minmax(220px,1fr));gap:12px}
.info-item{background:#ecf0f1;padding:12px;border-radius:6px}
.screenshot-box img{max-width:400px;border:2px solid #bdc3c7;border-radius:4px;margin-top:10px}
.no-screenshots{text-align:center;padding:40px;color:#7f8c8d;background:#ecf0f1;border-radius:8px}
.image-not-found{color:#e74c3c;background:#ffeaea;padding:10px;border-radius:4px;margin-top:8px}
.footer{text-align:center;color:#7f8c8d;margin-top:30px;padding-top:20px;border-top:1px solid #ecf0f1}
</style></head><body>
<h1>Vi SIM Toolkit Screenshot Report — ${escapeHtml(dialingNumber)}</h1>
<div class="info-box"><h2>Test Information</h2><div class="info-grid">
<div class="info-item"><strong>Device Number</strong><span>${escapeHtml(dialingNumber)}</span></div>
<div class="info-item"><strong>Test Start</strong><span>${fmt(start)}</span></div>
<div class="info-item"><strong>Test End</strong><span>${fmt(end)}</span></div>
<div class="info-item"><strong>Total Screenshots</strong><span>${shots.length}</span></div>
</div></div>
<div class="info-box"><h2>Screenshots Captured During Test</h2>${body}</div>
<div class="footer"><p>Generated by WebdriverIO + TypeScript Automation Framework</p>
<p>Report generated: ${new Date().toLocaleString()}</p></div>
</body></html>`;
}

function escapeHtml(s: string): string {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
}
